google-cloud-spanner/lib/google/cloud/spanner/transaction.rb

```diff
@@ -83,13 +84,21 @@ class Transaction
         def initialize
           @commit = Commit.new
           @seqno = 0
+
+          @mutex = Mutex.new
```

Comment:
We haven't done this very well in the existing code, but it's good practice to document the invariants enforced by mutexes because concurrency is hard on future maintainers.

Reply:
I documented the invariant at it's definition. Let me know if you prefer a more formal YARD style documentation (I avoided since it requires exposing the field through `attr_reader`).

Reply:
The documentation's location is fine; it's internal implementation documentation that doesn't need to be public. However, I was looking for something a bit more detailed on the actual invariant(s) being enforced, so that a future maintainer will know what kind of code needs to be protected by this mutex, and what they can assume when synchronizing on it. Especially because in this case there are multiple things being protected. My suggestion is something like this:

> This mutex protects two things: (1) the generation of sequence numbers, and (2) the creation of transactions. Specifically, `@seqno` is protected so it always reflects the last sequence number generated and provided to an operation in any thread. Any acquisition of a sequence number must be synchronized. Furthermore, `@grpc` is protected such that it either is `nil` if the transaction has not yet been created, or references the transaction resource if the transaction has been created. Any operation that could create a transaction must be synchronized, and any logic that depends on the state of transaction creation must also be synchronized.

Reply:
Thanks for the writeup :) I updated the file with the documentation.